prettier/src/utils/tokenize.ts

```diff
@@ -0,0 +1,160 @@
+    hublComment: "hublComment",
+    hublTag: "hublTag",
+    hublVariable: "hublVariable",
+  };
+
+  const tokenTypes = {
+    nestedHTML: {
+      placeholder: "npe$t_",
+    },
+    nestedStyleTag: {
+      placeholder: "/*styleblock$t*/",
+    },
+    jsonBlock: {
+      placeholder: "<!--placeholder-$t-->",
+      tokenReplacer: (match) => {
+        return `{% json_block %}${match}{% end_json_block %}`;
+      },
+    },
+    hublComment: { placeholder: "<!--$t-->" },
+    hublTag: {
+      placeholder: "<!--placeholder-$t-->",
+      tokenReplacer: (match) => {
+        return match.replace(LINE_BREAK_REGEX, " ");
+      },
+    },
+    hublVariable: {
+      placeholder: "<!--placeholder-$t-->",
+    },
+  };
+
+  const createToken = (sourceString, re, type) => {
+    const { tokenReplacer, placeholder } = tokenTypes[type];
+
+    return sourceString.replace(re, (token) => {
+      const existingToken = getTokenByValue(token, type);
+      if (existingToken) {
+        return existingToken;
+      }
+
+      tokenIndex++;
+      const placeholderKey = placeholder.replace("$t", tokenIndex);
+      const tokenValue = {
+        token: !!tokenReplacer ? tokenReplacer(token) : token,
+        type,
+      };
+
+      tokenPlaceholderMap.set(placeholderKey, tokenValue);
+
+      return placeholderKey;
+    });
+  };
+
+  // Create placeholders for tokens inside of <style> blocks
+  const nestedStyleTags = input.match(STYLE_BLOCK_WITH_HUBL_REGEX);
+  if (nestedStyleTags) {
+    input.match(STYLE_BLOCK_WITH_HUBL_REGEX).forEach((tag) => {
+      let newString = tag;
+
+      newString = createToken(
+        newString,
+        HUBL_TAG_REGEX,
+        TOKEN_TYPES.nestedStyleTag
+      );
+      newString = createToken(
+        newString,
+        HUBL_VARIABLE_REGEX,
+        TOKEN_TYPES.nestedStyleTag
+      );
+      newString = createToken(
+        newString,
+        HUBL_COMMENT_REGEX,
+        TOKEN_TYPES.nestedStyleTag
+      );
+
+      input = input.replace(tag, newString);
+    });
+  }
+
+  // Create placeholders for tokens within an html tag
+  const nestedHtmlTags = input.match(HTML_TAG_WITH_HUBL_TAG_REGEX);
+  if (nestedHtmlTags) {
+    nestedHtmlTags.forEach((tag) => {
+      let newString = tag;
+
+      newString = createToken(
+        newString,
+        HUBL_TAG_REGEX,
+        TOKEN_TYPES.nestedHTML
+      );
+      newString = createToken(
+        newString,
+        HUBL_VARIABLE_REGEX,
+        TOKEN_TYPES.nestedHTML
+      );
+
+      input = input.replace(tag, newString);
+    });
+  }
+
+  input = createToken(input, HUBL_COMMENT_REGEX, TOKEN_TYPES.hublComment);
+  input = createToken(input, JSON_BLOCK_REGEX, TOKEN_TYPES.jsonBlock);
+  input = createToken(input, HUBL_TAG_REGEX, TOKEN_TYPES.hublTag);
+  input = createToken(input, HUBL_VARIABLE_REGEX, TOKEN_TYPES.hublVariable);
+
+  // Reset index for next document
+  tokenIndex = 0;
+  return input;
+};
+export const untokenize = (input) => {
```

Comment:
This just grabs iterates over each placeholder and puts the hubl back where it belongs. Except for style blocks where we need to do a little more work because the HTML formatter adds extra spaces